src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -128,36 +128,67 @@ static async ValueTask WaitForConnectWithCancellation(AwaitableSocketAsyncEventA
 
         internal ValueTask ConnectAsync(IPAddress[] addresses, int port, CancellationToken cancellationToken)
         {
+            ThrowIfDisposed();
+
             if (addresses == null)
             {
                 throw new ArgumentNullException(nameof(addresses));
             }
+
             if (addresses.Length == 0)
             {
                 throw new ArgumentException(SR.net_invalidAddressList, nameof(addresses));
             }
 
-            return DoConnectAsync(addresses, port, cancellationToken);
-        }
+            if (!TcpValidationHelpers.ValidatePortNumber(port))
+            {
+                throw new ArgumentOutOfRangeException(nameof(port));
+            }
 
-        private async ValueTask DoConnectAsync(IPAddress[] addresses, int port, CancellationToken cancellationToken)
-        {
-            Exception? lastException = null;
-            foreach (IPAddress address in addresses)
+            if (_isListening)
             {
-                try
-                {
-                    await ConnectAsync(address, port, cancellationToken).ConfigureAwait(false);
-                    return;
-                }
-                catch (Exception ex) when (ex is not OperationCanceledException)
+                throw new InvalidOperationException(SR.net_sockets_mustnotlisten);
+            }
+
+            if (_isConnected)
+            {
+                throw new SocketException((int)SocketError.IsConnected);
+            }
+
+            ValidateForMultiConnect(isMultiEndpoint: false);
+
+            return Core(addresses, port, cancellationToken);
+
+            async ValueTask Core(IPAddress[] addresses, int port, CancellationToken cancellationToken)
+            {
+                Exception? lastException = null;
+                IPEndPoint? endPoint = null;
+                foreach (IPAddress address in addresses)
                 {
-                    lastException = ex;
+                    try
+                    {
+                        if (endPoint is null)
```

Comment:
I assume the point here is to avoid allocation when there are no addresses?

Seems like it would be clearer to just explicitly check for an empty array and then bail early from the method. That way endPoint doesn't have to be nullable.

Reply:
No,  it's to allocate the end point once with the first address and then reuse that same instance with each subsequent address.

Reply:
Yeah, sorry, makes sense.